sentinel-transport/sentinel-transport-common/src/main/java/com/alibaba/csp/sentinel/transport/message/AbstractHeartbeatMessage.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 1999-2018 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.csp.sentinel.transport.message;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import static com.alibaba.csp.sentinel.transport.message.HeartbeatMessageKeyConstants.*;
+import static com.alibaba.csp.sentinel.transport.message.HeartbeatMessageSuppliers.*;
+
+/**
+ * If you want to custom {@link HeartbeatMessage}, recommend that inherit this abstract class,
+ * <p>
+ * and use {@link #registerStaticInformation(String, String)} or {@link #registerDynamicInformation(String, Supplier)} to add the information you want.
+ *
+ * @author wxq
+ * @since 1.8.2
+ */
+public abstract class AbstractHeartbeatMessage implements HeartbeatMessage {
+
+    private final Map<String, String> staticMessage = new ConcurrentHashMap<>();
+
+    private final Map<String, Supplier<String>> dynamicMessage = new ConcurrentHashMap<>();
```

Comment:
Using `HashMap` maybe better, since the instance of `HeartbeatMessage` created by `SpiLoader` is singleton, no concurrent access.

Reply:
The refresh or register maybe concurrent, is it ok?

Reply:
Indeed, it's ok.